File: data_structures/fenwick_tree/matlab/test_fenwick_tree.m
% test_fenwick_tree.m  (1始まりインデックス)

ft = fenwick_tree('new', 6);
for i = 1:6
    ft = fenwick_tree(ft, 'update', i, i);
end
assert(fenwick_tree(ft, 'prefix_sum', 6) == 21);
assert(fenwick_tree(ft, 'range_sum', 2, 4) == 9);

data = [1, 3, 5, 7, 9, 11];
ft2 = fenwick_tree('from_data', data);
assert(fenwick_tree(ft2, 'prefix_sum', 6) == 36);
assert(fenwick_tree(ft2, 'range_sum', 2, 4) == 15);
ft2 = fenwick_tree(ft2, 'update', 3, 5);
assert(fenwick_tree(ft2, 'range_sum', 2, 4) == 20);

data3 = 1:100;
ft3 = fenwick_tree('from_data', data3);
assert(fenwick_tree(ft3, 'range_sum', 21, 50) == sum(data3(21:50)));
ft3 = fenwick_tree(ft3, 'update', 31, 900);
data3(31) = data3(31) + 900;
assert(fenwick_tree(ft3, 'range_sum', 21, 50) == sum(data3(21:50)));

try
    fenwick_tree(ft, 'update', 99, 0);
    error('expected error not thrown');
catch e
    assert(strcmp(e.identifier, 'fenwick_tree:outOfRange'));
end

disp('All tests passed.');

File: data_structures/fenwick_tree/python/test_fenwick_tree.py
import pytest
from fenwick_tree import FenwickTree


def test_manual_update():
    ft = FenwickTree(6)
    for i in range(6):
        ft.update(i, i + 1)
    assert ft.prefix_sum(6) == 21
    assert ft.range_sum(1, 4) == 9


def test_from_data():
    data = [1, 3, 5, 7, 9, 11]
    ft = FenwickTree(data)
    assert ft.prefix_sum(6) == 36
    assert ft.range_sum(1, 4) == 15
    ft.update(2, 5)
    assert ft.range_sum(1, 4) == 20


def test_brute_force():
    data = list(range(1, 101))
    ft = FenwickTree(data)
    assert ft.range_sum(20, 50) == sum(data[20:50])
    ft.update(30, 900)
    data[30] += 900
    assert ft.range_sum(20, 50) == sum(data[20:50])


def test_bad_index_raises():
    ft = FenwickTree(3)
    with pytest.raises(IndexError):
        ft.update(5, 1)
    with pytest.raises(IndexError):
        ft.range_sum(2, 1)


def test_float_elements():
    ft = FenwickTree(3)
    ft.update(0, 0.5)
    ft.update(1, 1.5)
    ft.update(2, 2.0)
    assert ft.prefix_sum(3) == 4.0
    assert ft.range_sum(1, 3) == 3.5
